### PR TITLE
Rails N + 2 queries

### PR DESCRIPTION
**Pull request by brunvez**
#### :link: Issue reference:

* Closes [Rails N + 2 queries](https://github.com/rootstrap/blog/issues/105)

---

#### Description:

Learn how N + 1 query optimizations might be hurting your performance and how to fix it.

---

#### :pushpin: Notes:

Should be understood by anyone who knows a bit of Rails

---

#### :file_folder: Categories

`Web/Backend/Rails/Performance`

---
#### :mag: Keywords:

- rails
- performance
- queries
- preload
- optimization
- activerecord

---

### CODE REVIEW

File: content/Rails N + 2 queries - Bruno Vezoli.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+Followed by your standard `irb` example:
+
+```ruby
+irb> posts = Post.all
+irb> all_comments = posts.map { |p| p.comments }
+```
+
+Which generates the following queries, where we can clearly see an N + 1 query since it needs to fetch comments for each
+post:
```
Comment (horacio) [inline content/Rails N + 2 queries - Bruno Vezoli.md:32]: As this is the gist of N+1, I'll complete this sentence with something emphasizing that.

```suggestion
Which generates the following queries where we can clearly see the infamous N+1 problem: one query for retrieving the posts and N additional queries for retrieving the comments, where N is the number of posts.
```

File: content/Rails N + 2 queries - Bruno Vezoli.md [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+Which in turn produces the following, much optimized, queries:
+
+```
+Post Load (0.5ms)  SELECT  "posts".* FROM "posts"
+Comment Load (1.4ms)  SELECT "comments".* FROM "comments" WHERE "comments"."post_id" IN ($1, $2, $3, $4, $5)  [["post_id", 1], ["post_id", 2], ["post_id", 3], ["post_id", 4], ["post_id", 5]]
+```
+
+And so we start using `#includes` on all our ActiveRecord queries and move on.
+
+### A little ways down the road
```
[the remark this answers is not in the document]
Comment (brunvez) [reply content/Rails N + 2 queries - Bruno Vezoli.md:59]: Yeah like time goes by and you will have to change code and this issue will reappear if you are not careful. The title is also a reference to How I met your mother 😅

File: content/Rails N + 2 queries - Bruno Vezoli.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+```ruby
+class Post < ApplicationRecord
+  has_many :comments
+end
+
+class Comment < ApplicationRecord
+  belongs_to :post
+end
+```
```
[the remark this answers is not in the document]
Comment (brunvez) [reply content/Rails N + 2 queries - Bruno Vezoli.md:22]: Hell yeah, I love Carbon. I use it on presentations, though I don't think it's possible to use it here right?

File: content/Rails N + 2 queries - Bruno Vezoli.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+And we get neat SQL queries once more
+
+```
+Post Load (0.4ms)  SELECT "posts".* FROM "posts"
+Comment Load (0.4ms)  SELECT "comments".* FROM "comments" WHERE "comments"."post_id" IN ($1, $2, $3, $4, $5) ORDER BY "comments"."likes" DESC  [["post_id", 1], ["post_id", 2], ["post_id", 3], ["post_id", 4], ["post_id", 5]]
+```
+
+Notice though that in this case we don't get the limit on the query and so Rails loads all these comments on memory and then
```
Comment (mariosouto) [inline content/Rails N + 2 queries - Bruno Vezoli.md:221]: `LIMIT`

File: content/Rails N + 2 queries - Bruno Vezoli.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+```
+Post Load (0.5ms)  SELECT  "posts".* FROM "posts"
+Comment Load (0.9ms)  SELECT "comments".* FROM "comments" WHERE "comments"."censored" = $1 AND "comments"."post_id" IN ($2, $3, $4, $5, $6)  [["censored", false], ["post_id", 1], ["post_id", 2], ["post_id", 3], ["post_id", 4], ["post_id", 5]]
+```
+
+### Caveats
+
+As with most performance optimizations you should really measure and evaluate the changes you are about to make. It doesn't
+really make sense to add an association to your models every time you want to preload, sometimes is better to have a small
```
Comment (andrea-fajardo) [inline content/Rails N + 2 queries - Bruno Vezoli.md:171]:
```suggestion
really make sense to add an association to your models every time you want to preload, sometimes it is better to have a small
```

File: content/Rails N + 2 queries - Bruno Vezoli.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+```
+
+We no longer get the warning and our queries are optimized:
+
+```
+Post Load (0.5ms)  SELECT  "posts".* FROM "posts"
+Comment Load (0.9ms)  SELECT "comments".* FROM "comments" WHERE "comments"."censored" = $1 AND "comments"."post_id" IN ($2, $3, $4, $5, $6)  [["censored", false], ["post_id", 1], ["post_id", 2], ["post_id", 3], ["post_id", 4], ["post_id", 5]]
+```
+
+### Caveats
```
Comment (andrea-fajardo) [inline content/Rails N + 2 queries - Bruno Vezoli.md:168]:
```suggestion
### Evaluating performance optimizations
```

**Pull request merged**